### PR TITLE
Allow multiple credentials secrets in single namespace

### PR DESCRIPTION
**Pull request by mshaposhnik**
## Summary

This PR allows the usage of mutiple build secrets per SCM type, host or repository.

It supports the `ssh-auth` and `basic-auth` type secrets.

Secrets must have the `appstudio.redhat.com/credentials: scm` and `appstudio.redhat.com/scm.host` labels, as well as might have an optional `appstudio.redhat.com/scm.repository` annotation value.

The best suitable one is selected, if none match then fallback is GitHub APP secret in the global namespace.

Known limitations:
 - GitHub dropped the login + password auth, seems only tokens are possible for now;
 - SSH key login isn't supported by any client, will need to find the way to perform authentication using it.

## Issue to Fix

https://issues.redhat.com/browse/SVPI-745
  
### Prerequisites:
 - GitHub organization
 - PAT with repo, user, org permissions;
 - Quay organization and repository with OAuth app enabled, known OAuth token for it.
  - Build-service fork and custom image. (typically, `mshaposh/build-controller:latest` is what in th PR)
 - Openshift cluster
 
### Installation
 - Get the latest version of `infra-deployments`
 - Create `hack/preview.env` from template. 
   Set the necessary fields:
   
   ```
   // local name of the fork fremote
   export MY_GIT_FORK_REMOTE=mshaposh
   ...
   // GH org and token
   export MY_GITHUB_ORG=mshaposhnik-org
   export MY_GITHUB_TOKEN=ghp_psj......
   ...
   // Quay org and token
   export IMAGE_CONTROLLER_QUAY_ORG=uy7c
   export IMAGE_CONTROLLER_QUAY_TOKEN=J8cM177j6......
   ...
   // Build service image and tag
   export BUILD_SERVICE_IMAGE_REPO=mshaposh/build-controller
   export BUILD_SERVICE_IMAGE_TAG=latest
   ```

 - Make sure `oc` is logged into a cluster, run from the infra-deployment project root:
  `./hack/bootstrap-cluster.sh preview --keycloak --toolchain`
  Make sure no fundamental errors during installlation.
  
 ### Testing
  - Create application from minimal template:
  ```
oc create -f - << EOF 
apiVersion: appstudio.redhat.com/v1alpha1
kind: Application
metadata:
  annotations:
  name: go-application-sample
  namespace: default
spec:
  appModelRepository:
    url: ''
  displayName: go-application-sample
  gitOpsRepository:
    url: ''
EOF
```
Wait it for resolve (must have status similar to this:)
```
 conditions:
    - lastTransitionTime: '2024-02-15T11:56:37Z'
      message: Application has been successfully created
      reason: OK
      status: 'True'
      type: Created
  devfile: |
    metadata:
      attributes:
        appModelRepository.context: /
        appModelRepository.url: https://github.com/mshaposhnik-org/go-application-sample-N6juw-marry-sign
        gitOpsRepository.context: ./
        gitOpsRepository.url: https://github.com/mshaposhnik-org/go-application-sample-N6juw-marry-sign
      name: go-application-sample
    schemaVersion: 2.2.0
```

- Create the SCM secret  (set or omit the repository annotation) :
```

apiVersion: v1
kind: Secret
metadata:
  namespace: default
  name: test-scm-secret
  labels:
    appstudio.redhat.com/credentials: scm
    appstudio.redhat.com/scm.host: github.com
  annotations:
    appstudio.redhat.com/scm.repository: mshaposhnik/*
type: kubernetes.io/basic-auth
stringData:
  password: Z2hw....
EOF
```

- Create component:
```
oc create -f - << EOF 
apiVersion: appstudio.redhat.com/v1alpha1
kind: Component
metadata:
  name: go-component-sample
  namespace: default
  annotations:
    build.appstudio.openshift.io/request: "configure-pac"
    image.redhat.com/generate: '{"visibility": "public"}'
spec:
  componentName: go-sample-component-name
  application: go-application-sample
  source:
    git:
      url: https://github.com/mshaposhnik/devfile-sample-go-basic.git
EOF
      
```
It must resolve status:
```
status:
  conditions:
    - lastTransitionTime: '2024-02-15T12:36:44Z'
      message: GitOps resource generated successfully
      reason: OK
      status: 'True'
      type: GitOpsResourcesGenerated
    - lastTransitionTime: '2024-02-15T12:36:44Z'
      message: Component has been successfully created
      reason: OK
      status: 'True'
      type: Created
```

PR to repository must be submitted.

### CODE REVIEW

File: controllers/component_build_controller_pac.go [hunk 1]
```diff
@@ -462,54 +466,150 @@ func (r *ComponentBuildReconciler) UndoPaCProvisionForComponent(ctx context.Cont
...
+		}
+
+		r.EventRecorder.Event(pacSecret, "Warning", "PaCSecretNotFound", err.Error())
+		// Do not trigger a new reconcile. The PaC secret must be created first.
+		return nil, boerrors.NewBuildOpError(boerrors.EPaCSecretNotFound, fmt.Errorf(" Pipelines as Code secret not found in %s ", globalPaCSecretKey.Namespace))
+	}
+	// There is some secret but doesn't look like a GH App secret
+	if !gitops.IsPaCApplicationConfigured("github", pacSecret.Data) {
+		r.EventRecorder.Event(pacSecret, "Warning", "PaCSecretNotFound", "PaC secret is not configured for GitHub App")
+		return nil, boerrors.NewBuildOpError(boerrors.EPaCSecretNotFound, fmt.Errorf(" Pipelines as Code secret not found in %s ", globalPaCSecretKey.Namespace))
```
Comment (mmorhun) [inline controllers/component_build_controller_pac.go]: I think that it's not a not found error.

File: controllers/component_build_controller_pac.go [hunk 2]
```diff
@@ -462,54 +466,145 @@ func (r *ComponentBuildReconciler) UndoPaCProvisionForComponent(ctx context.Cont
...
-			r.EventRecorder.Event(&pacSecret, "Warning", "PaCSecretNotFound", err.Error())
-			// Do not trigger a new reconcile. The PaC secret must be created first.
-			return nil, boerrors.NewBuildOpError(boerrors.EPaCSecretNotFound,
-				fmt.Errorf(" Pipelines as Code secret not found in %s namespace nor in %s", pacSecretKey.Namespace, globalPaCSecretKey.Namespace))
+	// find the best matching secret, starting from SSH type
+	if sshSecret := bestMatchingSecret(ctx, cRepo, sshSecrets); sshSecret != nil {
+		return sshSecret, nil
+	} else if basicSecret := bestMatchingSecret(ctx, cRepo, basicSecrets); basicSecret != nil {
+		return basicSecret, nil
+	} else {
```
Comment (mmorhun) [inline controllers/component_build_controller_pac.go:519]: We don't need all those `else`s. Just `if` + `return`

File: controllers/component_build_controller_pac.go [hunk 3]
```diff
@@ -462,54 +466,145 @@ func (r *ComponentBuildReconciler) UndoPaCProvisionForComponent(ctx context.Cont
...
-				ObjectMeta: metav1.ObjectMeta{
-					Name:      pacSecretKey.Name,
-					Namespace: pacSecretKey.Namespace,
-					Labels: map[string]string{
-						PartOfLabelName: PartOfAppStudioLabelValue,
-					},
-				},
-				Data: pacSecret.Data,
+// finds the best matching secret for the given repository
+func bestMatchingSecret(_ context.Context, cRepo string, secrets []corev1.Secret) *corev1.Secret {
```
Comment (mmorhun) [inline controllers/component_build_controller_pac.go:530]: Why do we need to pass and the ignore context?

File: controllers/component_build_controller_pac.go [hunk 4]
```diff
@@ -462,54 +466,145 @@ func (r *ComponentBuildReconciler) UndoPaCProvisionForComponent(ctx context.Cont
...
+
+	// secrets without repository annotation
+	var hostOnlySecrets []corev1.Secret
+
+	// map of secret idx and its best path intersections count
```
Comment (mmorhun) [inline controllers/component_build_controller_pac.go:535]: Comment is not clear to me.
 - does idx stand for index? Please use full word in comment.
 - what is "best path intersections count"?

File: controllers/component_build_controller_pac.go [hunk 5]
```diff
@@ -462,54 +466,145 @@ func (r *ComponentBuildReconciler) UndoPaCProvisionForComponent(ctx context.Cont
...
+	var potentialMatches = make(map[int]int, len(secrets))
+
+	for idx, secret := range secrets {
```
Comment (mmorhun) [inline controllers/component_build_controller_pac.go:538]: I'd prefer `i` or `index`

File: controllers/component_build_controller_pac.go [hunk 6]
```diff
@@ -462,54 +466,145 @@ func (r *ComponentBuildReconciler) UndoPaCProvisionForComponent(ctx context.Cont
...
+		annValue, ok := secret.Annotations[scmSecretRepositoryAnnotation]
```
Comment (mmorhun) [inline controllers/component_build_controller_pac.go:539]: `annValue` is not common abbreviation, please use full work for clarity

File: controllers/component_build_controller_pac.go [hunk 7]
```diff
@@ -462,54 +466,145 @@ func (r *ComponentBuildReconciler) UndoPaCProvisionForComponent(ctx context.Cont
...
+		if !ok || annValue == "" {
```
Comment (mmorhun) [inline controllers/component_build_controller_pac.go:540]: Does it make sense to `ok` -> `exists`?

File: controllers/component_build_controller_pac.go [hunk 8]
```diff
@@ -462,54 +466,145 @@ func (r *ComponentBuildReconciler) UndoPaCProvisionForComponent(ctx context.Cont
...
+			hostOnlySecrets = append(hostOnlySecrets, secret)
+			continue
+		}
+		repos := strings.Split(annValue, ",")
+
+		// Direct repository match, return secret
+		if slices.Contains(repos, cRepo) {
+			return &secret
+		}
+		// No direct match, check for wildcard match
```
Comment (mmorhun) [inline controllers/component_build_controller_pac.go:550]: An example of wildcard match would make it more clear.

File: controllers/component_build_controller_pac.go [hunk 9]
```diff
@@ -462,54 +466,145 @@ func (r *ComponentBuildReconciler) UndoPaCProvisionForComponent(ctx context.Cont
...
-				TypeMeta: pacSecret.TypeMeta,
-				ObjectMeta: metav1.ObjectMeta{
-					Name:      pacSecretKey.Name,
-					Namespace: pacSecretKey.Namespace,
-					Labels: map[string]string{
-						PartOfLabelName: PartOfAppStudioLabelValue,
-					},
-				},
-				Data: pacSecret.Data,
+// finds the best matching secret for the given repository
```
Comment (mmorhun) [inline controllers/component_build_controller_pac.go:529]: Please expand the function description, so one can understand what would be returned without reading the code.
Examples appreciated.

File: controllers/component_build_controller_pac.go [hunk 10]
```diff
@@ -462,54 +466,145 @@ func (r *ComponentBuildReconciler) UndoPaCProvisionForComponent(ctx context.Cont
...
 			}
-			if err := r.Client.Create(ctx, localPaCSecret); err != nil {
-				return nil, fmt.Errorf("failed to create local PaC configuration secret: %w", err)
+		}
+	}
+	if len(potentialMatches) == 0 && len(hostOnlySecrets) == 0 {
+		return nil // Nothing matched
+	} else if len(potentialMatches) == 0 && len(hostOnlySecrets) > 0 {
+		return &hostOnlySecrets[0] // Return first host-only secret
+	} else {
```
Comment (mmorhun) [inline controllers/component_build_controller_pac.go:568]: `else`s not needed here

File: controllers/component_build_controller_pac.go [hunk 11]
```diff
@@ -462,54 +466,145 @@ func (r *ComponentBuildReconciler) UndoPaCProvisionForComponent(ctx context.Cont
...
+				bestCount = count
+				bestIdx = idx
 			}
 		}
+		return &secrets[bestIdx]
 	}
+}
 
-	return &pacSecret, nil
+// slicesIntersection returns the number of intersections between two slices with respect to the order of elements
```
Comment (mmorhun) [inline controllers/component_build_controller_pac.go:581]: Could you make the description more clear and add example(s)?

File: pkg/git/gitproviderfactory/gitproviderfactory.go [hunk 12]
```diff
@@ -17,22 +17,22 @@ limitations under the License.
...
 	"fmt"
-	"strconv"
-	"strings"
-
 	"github.com/redhat-appstudio/application-service/gitops"
 	"github.com/redhat-appstudio/build-service/pkg/boerrors"
 	"github.com/redhat-appstudio/build-service/pkg/git/github"
 	"github.com/redhat-appstudio/build-service/pkg/git/gitlab"
 	"github.com/redhat-appstudio/build-service/pkg/git/gitprovider"
+	corev1 "k8s.io/api/core/v1"
```
Comment (mmorhun) [inline pkg/git/gitproviderfactory/gitproviderfactory.go:27]: Can we avoid bringing k8s dependencies into git packages?

File: controllers/component_build_controller_pac.go [hunk 13]
```diff
@@ -75,6 +75,10 @@ const (
 	appstudioWorkspaceNameLabel      = "appstudio.redhat.com/workspace_name"
 	pacCustomParamAppstudioWorkspace = "appstudio_workspace"
 
+	scmSecretMarkerLabel          = "appstudio.redhat.com/credentials"
```
[the remark this answers is not in the document]
Comment (mshaposhnik) [reply controllers/component_build_controller_pac.go:78]: As we agreed, it's a `appstudio.redhat.com/credentials=scm` label

File: controllers/component_build_controller_pac.go (hunk 2 above)
Comment (mshaposhnik) [reply controllers/component_build_controller_pac.go]: ok

File: controllers/component_build_controller_pac.go (hunk 3 above)
Comment (mshaposhnik) [reply controllers/component_build_controller_pac.go]: ok

File: pkg/git/gitproviderfactory/gitproviderfactory.go (hunk 12 above)
Comment (mshaposhnik) [reply pkg/git/gitproviderfactory/gitproviderfactory.go:27]: is that a big concern? Im afraid i cannot avoid that w/out adding some degree of mess :(

Comment (mmorhun) [reply pkg/git/gitproviderfactory/gitproviderfactory.go:27]: @mshaposhnik, the idea is to have git packages independent from the main code, so they can be reused from other places like GitHub actions. While creating a map is not a big deal, building k8s secret object and kubernetes dependencies looks like overhead.
I think we can easily remove that dependency, by just passing a map with credentials as it was done before. All the validations are already happened, so we can be sure we have a valid structure and can easily guess auth type based on the fields in the map.

Comment (mshaposhnik) [reply pkg/git/gitproviderfactory/gitproviderfactory.go]: ok, i refactored it

File: pkg/git/gitproviderfactory/gitproviderfactory.go [hunk 14]
```diff
@@ -50,28 +49,50 @@ type GitClientConfig struct {
...
-	if !isAppUsed {
-		accessToken = strings.TrimSpace(string(config[gitops.GetProviderTokenKey(gitProvider)]))
-	}
+	isAppUsed := gitops.IsPaCApplicationConfigured(gitProvider, secretData)
 
 	switch gitProvider {
 	case "github":
 		if !isAppUsed {
-			return github.NewGithubClient(accessToken), nil
+			if usernameExists && passwordExists {
```
Comment (mmorhun) [inline pkg/git/gitproviderfactory/gitproviderfactory.go:62]: I think we can have `if passwordExists {` and then analyze username. That would remove duplication on password decoding. But up to you.

**Review by mmorhun — APPROVED** [review]

**Review by tkdchen — COMMENTED** [review]
LGTM. There is Envtest failure.

**Review by tkdchen — APPROVED** [review]

**Pull request merged**